src/vs/workbench/contrib/debug/common/replModel.ts

```diff
@@ -6,15 +6,107 @@
 import { IReplElement, IStackFrame, IExpression, IReplElementSource, IDebugSession } from 'vs/workbench/contrib/debug/common/debug';
-import { Expression, SimpleReplElement, RawObjectReplElement } from 'vs/workbench/contrib/debug/common/debugModel';
-import { isUndefinedOrNull, isObject } from 'vs/base/common/types';
+import { ExpressionContainer } from 'vs/workbench/contrib/debug/common/debugModel';
+import { isString, isUndefinedOrNull, isObject } from 'vs/base/common/types';
 import { basenameOrAuthority } from 'vs/base/common/resources';
 import { URI } from 'vs/base/common/uri';
 import { endsWith } from 'vs/base/common/strings';
+import { generateUuid } from 'vs/base/common/uuid';
 
 const MAX_REPL_LENGTH = 10000;
 let topReplElementCounter = 0;
 
+export class SimpleReplElement implements IReplElement {
+	constructor(
+		private id: string,
+		public value: string,
+		public severity: severity,
+		public sourceData?: IReplElementSource,
+	) { }
+
+	toString(): string {
+		return this.value;
+	}
+
+	getId(): string {
+		return this.id;
+	}
+}
+
+export class RawObjectReplElement implements IExpression {
+
+	private static readonly MAX_CHILDREN = 1000; // upper bound of children per value
+
+	constructor(private id: string, public name: string, public valueObj: any, public sourceData?: IReplElementSource, public annotation?: string) { }
+
+	getId(): string {
+		return this.id;
+	}
+
+	get value(): string {
+		if (this.valueObj === null) {
+			return 'null';
+		} else if (Array.isArray(this.valueObj)) {
+			return `Array[${this.valueObj.length}]`;
+		} else if (isObject(this.valueObj)) {
+			return 'Object';
+		} else if (isString(this.valueObj)) {
+			return `"${this.valueObj}"`;
+		}
+
+		return String(this.valueObj) || '';
+	}
+
+	get hasChildren(): boolean {
+		return (Array.isArray(this.valueObj) && this.valueObj.length > 0) || (isObject(this.valueObj) && Object.getOwnPropertyNames(this.valueObj).length > 0);
+	}
+
+	getChildren(): Promise<IExpression[]> {
+		let result: IExpression[] = [];
+		if (Array.isArray(this.valueObj)) {
+			result = (<any[]>this.valueObj).slice(0, RawObjectReplElement.MAX_CHILDREN)
+				.map((v, index) => new RawObjectReplElement(`${this.id}:${index}`, String(index), v));
+		} else if (isObject(this.valueObj)) {
+			result = Object.getOwnPropertyNames(this.valueObj).slice(0, RawObjectReplElement.MAX_CHILDREN)
+				.map((key, index) => new RawObjectReplElement(`${this.id}:${index}`, key, this.valueObj[key]));
+		}
+
+		return Promise.resolve(result);
+	}
+
+	toString(): string {
+		return `${this.name}\n${this.value}`;
+	}
+}
+
+export class ReplEvaluationInput implements IReplElement {
+	private id: string;
+
+	constructor(public value: string) {
+		this.id = generateUuid();
+	}
+
+	toString(): string {
+		return this.value;
+	}
+
+	getId(): string {
+		return this.id;
+	}
+}
+
+export class ReplEvaluationResult extends ExpressionContainer implements IReplElement, IExpression {
```

Comment:
I do not think `ReplEvaluationResult` should implement `IExpression` and it should not have a `public name` which we do not use anywhere.
I see that you are passing `ReplEvaluationResult` to `renderExpressionValue`, but can we pass just it's `value` instead.
I think we get rid of the `name` it will get clearer that the ReplEvaluationResult is just the result

I might be missing something here, but idealy we can simplify this.